### PR TITLE
[MI-2466] Added testcases for run stage job request feature

### PR DESCRIPTION
**Pull request by raghavaggarwal2308**

### CODE REVIEW

File: server/plugin/api_test.go [hunk 1]
```diff
@@ -960,3 +961,172 @@ func TestGetUserChannelsForTeam(t *testing.T) {
...
+				"channel_id": "mockChannelID",
+				"context": {
+				  "approvalId": "mockApproverID",
+				  "organization": "mockOrganization",
+				  "projectID": "mockProjectID",
+				  "requestType": "mockRequestType"
+				}
+			  }`,
+			updatePipelineRunApprovalRequestStatus: http.StatusOK,
+			updatePipelineRunApprovalPostError:     errors.New("approved/rejected request successfully but failed to update post"),
```
Comment (Kshitij-Katiyar) [inline server/plugin/api_test.go:1011]: same

File: server/plugin/api_test.go [hunk 2]
```diff
@@ -960,3 +961,172 @@ func TestGetUserChannelsForTeam(t *testing.T) {
...
+			monkey.PatchInstanceMethod(reflect.TypeOf(p), "UpdatePipelineRunApprovalPost", func(_ *Plugin, _ []*serializers.ApprovalStep, _ int, _, _, _ string) error {
+				return testCase.updatePipelineRunApprovalPostError
+			})
+
+			monkey.PatchInstanceMethod(reflect.TypeOf(p), "DM", func(_ *Plugin, _, _ string, _ bool, _ ...interface{}) (string, error) {
+				return "", nil
+			})
+
+			req := httptest.NewRequest(http.MethodGet, "/mockPath", bytes.NewBufferString(testCase.body))
+			req.Header.Add(constants.HeaderMattermostUserID, "test-userID")
```
Comment (Kshitij-Katiyar) [inline server/plugin/api_test.go:1124]: make a constant for this ID as I think this is going to be used many times

File: server/plugin/client_test.go [hunk 3]
```diff
@@ -317,6 +317,80 @@ func TestCall(t *testing.T) {
...
+	defer monkey.UnpatchAll()
+	p := setupTestPlugin(&plugintest.API{})
+	for _, testCase := range []struct {
+		description string
+		err         error
+		statusCode  int
+	}{
+		{
+			description: "UpdatePipelineRunApprovalRequest: valid",
+			err:         nil,
```
Comment (ayusht2810) [inline server/plugin/client_test.go:330]:
```suggestion
```

Comment (ayusht2810) [reply server/plugin/client_test.go:330]: check everywhere

File: server/plugin/client_test.go [hunk 4]
```diff
@@ -317,6 +317,80 @@ func TestCall(t *testing.T) {
...
+			statusCode:  http.StatusOK,
+		},
+		{
+			description: "UpdatePipelineRunApprovalRequest: with error",
+			err:         errors.New("mock-error"),
```
Comment (ayusht2810) [inline server/plugin/client_test.go:335]: change error and check everywhere

File: server/plugin/client_test.go [hunk 5]
```diff
@@ -317,6 +317,80 @@ func TestCall(t *testing.T) {
...
+	} {
+		t.Run(testCase.description, func(t *testing.T) {
+			monkey.PatchInstanceMethod(reflect.TypeOf(&client{}), "Call", func(_ *client, basePath, method, path, contentType, mattermostUserID string, inBody io.Reader, out interface{}, formValues url.Values) (responseData []byte, statusCode int, err error) {
+				return nil, testCase.statusCode, testCase.err
+			})
+
+			_, statusCode, err := p.Client.UpdatePipelineRunApprovalRequest([]*serializers.PipelineApproveRequest{}, "mockProjectID", "mockMattermostUSerID", "mockApprovalID")
+
+			if testCase.err != nil {
+				assert.Error(t, err)
```
Comment (ayusht2810) [inline server/plugin/client_test.go:347]: check for error value if possible and check everywhere

File: server/plugin/api_test.go [hunk 6]
```diff
@@ -960,3 +961,172 @@ func TestGetUserChannelsForTeam(t *testing.T) {
...
+				  "projectID": "mockProjectID",
+				  "requestType": "mockRequestType"
+				}
+			  }`,
+			updatePipelineRunApprovalRequestStatus: http.StatusInternalServerError,
+			updatePipelineRunApprovalRequestError:  errors.New("not permitted to complete approval"),
+			statusCode:                             http.StatusOK,
+		},
+		{
+			description: "HandlePipelineApproveOrRejectRunRequest: failed to approve/reject request and update the post",
```
Comment (ayusht2810) [inline server/plugin/api_test.go:1031]:
```suggestion
			description: "HandlePipelineApproveOrRejectRunRequest: failed to approve/reject the request and update the post",
```

Comment (ayusht2810) [reply server/plugin/api_test.go:1031]: check everywhere

File: server/plugin/api_test.go (hunk 2 above)
Comment (raghavaggarwal2308) [reply server/plugin/api_test.go:1124]: Created a ticket for this. I am doing the same thing across multiple PRs. Once all the PRs are merged, I will create a separate PR for such small enhancements.
https://brightscout.atlassian.net/browse/MI-2511?atlOrigin=eyJpIjoiNWFlOTMxODJjYWM4NDgwYWJiNjE2N2MxMDdjZDRkNjAiLCJwIjoiaiJ9

File: server/plugin/client_test.go (hunk 4 above)
Comment (raghavaggarwal2308) [reply server/plugin/client_test.go:335]: Created a ticket for this. I am doing the same thing across multiple PRs. Once all the PRs are merged, I will create a separate PR for such small enhancements.
https://brightscout.atlassian.net/browse/MI-2511?atlOrigin=eyJpIjoiNWFlOTMxODJjYWM4NDgwYWJiNjE2N2MxMDdjZDRkNjAiLCJwIjoiaiJ9

File: server/plugin/api_test.go [hunk 7]
```diff
@@ -960,3 +961,172 @@ func TestGetUserChannelsForTeam(t *testing.T) {
...
+				  "projectID": "mockProjectID",
+				  "requestType": "mockRequestType"
+				}
+			  }`,
+			updatePipelineRunApprovalRequestStatus: http.StatusOK,
+			statusCode:                             http.StatusOK,
+			getRunApprovalDetailsStatus:            http.StatusOK,
+		},
+		{
+			description: "HandlePipelineApproveOrRejectRunRequest: approved/rejected request successfully but failed to update post",
```
Comment (Kshitij-Katiyar) [inline server/plugin/api_test.go:999]:
```suggestion
			description: "HandlePipelineApproveOrRejectRunRequest: approved/rejected request successfully but failed to update the post",
```

Comment (raghavaggarwal2308) [reply server/plugin/api_test.go:999]: discussed

**Review by ayusht2810 — APPROVED** [review]

File: server/plugin/api_test.go [hunk 8]
```diff
@@ -960,3 +961,172 @@ func TestGetUserChannelsForTeam(t *testing.T) {
...
+				"channel_id": "mockChannelID",
+				"context": {
+				  "approvalId": "mockApproverID",
+				  "organization": "mockOrganization",
+				  "projectID": "mockProjectID",
+				  "requestType": "mockRequestType"
+				}
+			  }`,
+			updatePipelineRunApprovalRequestStatus: http.StatusInternalServerError,
+			updatePipelineRunApprovalRequestError:  errors.New("not permitted to complete approval"),
```
Comment (Kshitij-Katiyar) [inline server/plugin/api_test.go:1027]:
```suggestion
			updatePipelineRunApprovalRequestError:  errors.New("not permitted to complete the approval"),
```

File: server/plugin/api_test.go [hunk 9]
```diff
@@ -960,3 +961,172 @@ func TestGetUserChannelsForTeam(t *testing.T) {
...
```
Comment (Kshitij-Katiyar) [inline server/plugin/api_test.go:1060]: same

File: server/plugin/api_test.go [hunk 10]
```diff
@@ -960,3 +961,172 @@ func TestGetUserChannelsForTeam(t *testing.T) {
...
+			if !testCase.isPayloadInvalid {
+				mockedClient.EXPECT().UpdatePipelineRunApprovalRequest(gomock.Any(), "mockOrganization", "mockProjectID", "test-userID").Return(&serializers.PipelineRunApproveResponse{
+					Value: []*serializers.PipelineRunResponseValue{
+						{},
+					},
+				}, testCase.updatePipelineRunApprovalRequestStatus, testCase.updatePipelineRunApprovalRequestError)
+			}
+
+			if testCase.updatePipelineRunApprovalRequestStatus == http.StatusInternalServerError {
+				mockedClient.EXPECT().GetRunApprovalDetails("mockOrganization", "mockProjectID", "test-userID", "mockApproverID").Return(&serializers.PipelineRunApprovalDetails{}, testCase.getRunApprovalDetailsStatus, testCase.getRunApprovalDetailsError)
```
Comment (Kshitij-Katiyar) [inline server/plugin/api_test.go:1112]: use the constants created in other branches here as well

Comment (raghavaggarwal2308) [reply server/plugin/api_test.go:1112]: Created a ticket for this. I am doing the same thing across multiple PRs. Once all the PRs are merged, I will create a separate PR for such small enhancements.
https://brightscout.atlassian.net/browse/MI-2511?atlOrigin=eyJpIjoiNWFlOTMxODJjYWM4NDgwYWJiNjE2N2MxMDdjZDRkNjAiLCJwIjoiaiJ9

File: server/plugin/api_test.go (hunk 8 above)
Comment (raghavaggarwal2308) [reply server/plugin/api_test.go:1027]: We cannot change this error. It is returned by Azure API

**Review by Kshitij-Katiyar — APPROVED** [review]

**Pull request merged**